sdk/core/Azure.Core/src/Shared/LightweightPkcs8Decoder.cs

```diff
@@ -31,6 +32,160 @@ internal class LightweightPkcs8Decoder
             0x05, 0x00,
         };
 
+        internal static byte[] ReadBitString(byte[] data, ref int offset)
+        {
+            // Adapted from https://github.com/dotnet/runtime/blob/be74b4bd/src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.BitString.cs#L156
+
+            if (data[offset++] != 0x03)
+            {
+                throw new InvalidDataException("Invalid PKCS#8 Data");
+            }
+
+            int length = ReadLength(data, ref offset);
+            if (length == 0)
+            {
+                throw new InvalidDataException("Invalid PKCS#8 Data");
+            }
+
+            int unusedBitCount = data[offset++];
+            if (unusedBitCount > 7)
+            {
+                throw new InvalidDataException("Invalid PKCS#8 Data");
+            }
+
+            Span<byte> span = data.AsSpan(offset, length - 1);
+
+            // Build a mask for the bits that are used so the normalized value can be computed
+            //
+            // If 3 bits are "unused" then build a mask for them to check for 0.
+            // -1 << 3 => 0b1111_1111 << 3 => 0b1111_1000
+            int mask = -1 << unusedBitCount;
+            byte lastByte = span[span.Length - 1];
+            byte maskedByte = (byte)(lastByte & mask);
+
+            byte[] ret = new byte[span.Length];
+
+            Buffer.BlockCopy(data, offset, ret, 0, span.Length);
+            ret[span.Length - 1] = maskedByte;
+
+            offset += span.Length;
+
+            return ret;
+        }
+
+        internal static string ReadObjectIdentifier(byte[] data, ref int offset)
+        {
+            // Adapted from https://github.com/dotnet/runtime/blob/be74b4bd/src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnDecoder.Oid.cs#L175
+
+            if (data[offset++] != 0x06)
+            {
+                throw new InvalidDataException("Invalid PKCS#8 Data");
+            }
+
+            int length = ReadLength(data, ref offset);
+
+            StringBuilder ret = new StringBuilder();
+            for (int i = offset; i < offset + length; i++)
+            {
+                byte val = data[i];
+
+                if (i == offset)
+                {
+                    byte first;
+                    if (val < 40)
+                    {
+                        first = 0;
+                    }
+                    else if (val < 80)
+                    {
+                        first = 1;
+                        val -= 40;
+                    }
+                    else
+                    {
+                        first = 2;
+                        val -= 80;
+                    }
+
+                    ret.Append(first).Append('.').Append(val);
```

Comment:
I think everyone makes this bug the first time around.

arcs 0 and 1 are both restricted to values 0..39 on the first sub-arc.  arc 2 isn't.  `2.999` is a legal OID, with an encoding of `06028837`.

You probably won't need to make it actually work for your limited needs here, but if you're not going to support it properly you probably want to throw in the else block if `val` had the high bit set (indicating a multi-byte value), which means you'll support 2.0 through 2.47.

According to oid-info.com, there are 5 real arcs (and the 2.999 example ID) that you won't support: 2.48-2.52.  http://oid-info.com/cgi-bin/display?oid=2&submit=Tree+display

Reply:
With limited support anyway, I'll add the `throw` and applicable test. Hopefully we'll be able to switch to the `AsnReader` soon-ish.